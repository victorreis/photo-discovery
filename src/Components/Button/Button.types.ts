import { TestProps } from '../../Config/Tests/Test.types';

export const availableButtonTypes = ['button', 'submit'] as const;
export type ButtonType = typeof availableButtonTypes[number];

export interface RequiredButtonProps {
  children: React.ReactNode;
  onClick: () => void;
}

export interface DefaultButtonProps {
  type?: ButtonType;
}

export interface OptionalButtonProps {
  style?: React.CSSProperties;
}

export type ButtonProps = RequiredButtonProps &
  DefaultButtonProps &
  OptionalButtonProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLButtonElement>, 'children' | 'onClick'>;

export type ButtonStyleProps = Required<DefaultButtonProps>;
