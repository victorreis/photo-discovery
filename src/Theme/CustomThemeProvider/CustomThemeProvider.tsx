import React, { useMemo, useState } from 'react';

import { ThemeProvider } from 'styled-components';

import {
  lightTheme,
  darkTheme,
  futuristicTheme,
  deepSpaceTheme,
} from '../Modes';
import { Mode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<Mode | 'default', CustomThemeType> = {
  default: deepSpaceTheme,
  light: lightTheme,
  dark: darkTheme,
  futuristic: futuristicTheme,
  deepSpace: deepSpaceTheme,
};

export const ThemeContext = React.createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<CustomThemeProviderProps> = {
  testID: 'ThemeContextProvider',
  themeName: 'deepSpace',
};

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: Mode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemeContext.Provider>
  );
};
