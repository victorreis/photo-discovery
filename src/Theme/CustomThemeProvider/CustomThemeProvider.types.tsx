import { CustomThemeType, Mode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: Mode) => void;
};

export type CustomThemeProviderProps = { testID?: string; themeName?: Mode };
