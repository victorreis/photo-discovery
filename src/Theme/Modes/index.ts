export { darkTheme } from './Dark.theme';
export { deepSpaceTheme } from './DeepSpace.theme';
export { futuristicTheme } from './Futuristic.theme';
export { lightTheme } from './Light.theme';
