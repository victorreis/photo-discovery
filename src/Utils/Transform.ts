export const hexToRgb = (hex: string) => {
  // Expand shorthand form (e.g. "03F") to full form (e.g. "0033FF")
  const shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
  const formatedHex = hex.replace(
    shorthandRegex,
    (_m, r: string, g: string, b: string) => r + r + g + g + b + b
  );

  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(formatedHex);
  return result
    ? `rgb(${parseInt(result[1] || '0', 16)}, ${parseInt(
        result[1] || '0',
        16
      )}, ${parseInt(result[1] || '0', 16)})`
    : 'rgb(0,0,0)';
};
