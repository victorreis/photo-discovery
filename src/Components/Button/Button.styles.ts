import styled from 'styled-components';

import { toPx } from '../../Utils/Transform';
import { ButtonStyleProps } from './Button.types';

export const ButtonContainer = styled.button<ButtonStyleProps>`
  cursor: pointer;
  height: 3rem;
  margin: 0.5rem;
  border: 0;
  letter-spacing: 0.2rem;
  text-transform: uppercase;
  align-items: center;

  ${({ theme }) => theme.typographies.button};
  background-color: ${({ theme }) => theme.colors.background.default.darkest};
  color: ${({ theme }) => theme.colors.background.default.lightest};
  border-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};

  box-shadow: inset 0 0 1em ${({ theme }) => theme.colors.main.effect.normal}55;

  &:hover {
    box-shadow: inset 0 0 1em ${({ theme }) => theme.colors.main.effect.dark},
      0 0 0.2em ${({ theme }) => theme.colors.main.effect.dark};
  }
`;
