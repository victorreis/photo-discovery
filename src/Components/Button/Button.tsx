import { TestProps } from '../../Config/Tests/Test.types';
import { ButtonContainer } from './Button.styles';
import { ButtonProps, DefaultButtonProps } from './Button.types';

export const buttonDefaults: Required<DefaultButtonProps> &
  Required<TestProps> = {
  testID: 'Button',
  type: 'button',
};

export const Button: React.FC<ButtonProps> = (props): JSX.Element => {
  const {
    testID = buttonDefaults.testID,
    children,
    type = buttonDefaults.type,
    onClick,
    ...others
  } = props;

  const handleClick = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    event.stopPropagation();
    if (onClick) {
      onClick();
    }
  };

  return (
    <ButtonContainer
      data-testid={testID}
      onClick={handleClick}
      type={type}
      {...others}
    >
      {children}
    </ButtonContainer>
  );
};
