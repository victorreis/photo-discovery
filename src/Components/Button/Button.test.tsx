import {
  fireEvent,
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb, toPx } from '../../Utils/Transform';
import { Button, buttonDefaults } from './Button';
import { RequiredButtonProps, ButtonProps } from './Button.types';

describe(`Button component tests`, () => {
  const text = 'text';
  const onClick = jest.fn();

  const requiredProps: RequiredButtonProps = {
    children: text,
    onClick,
  };

  const setup = (props?: ButtonProps) => {
    const renderRTR = () =>
      renderRTRCreator<ButtonProps>(Button, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<ButtonProps>(Button, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe(`behavior tests`, () => {
    it(`should render the component`, () => {
      setup().renderJestDom();
      const testInstance = screen.getByTestId(buttonDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      setup().renderJestDom();
      const element = screen.getByText(text);

      expect(element).toBeInTheDocument();
    });

    it(`should call the onCLick callback when the button is clicked`, () => {
      setup().renderJestDom();
      const testInstance = screen.getByTestId(buttonDefaults.testID);

      fireEvent.click(testInstance);
      expect(onClick).toHaveBeenCalled();
    });
  });

  describe(`style tests`, () => {
    it(`should have style the correct styles`, () => {
      setup().renderJestDom();
      const container = screen.getByTestId(buttonDefaults.testID);

      expect(container).toHaveStyle({
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.darkest
        ),
        borderRadius: toPx(themes.default.borders.radius.LG),
        color: hexToRgb(themes.default.colors.background.default.lightest),
        ...themes.default.typographies.button,
      });
    });
  });

  describe(`snapshot tests`, () => {
    it(`should render correctly`, () => {
      const generatedJson = setup().renderRTR().toJSON();
      expect(generatedJson).toMatchSnapshot();
    });
  });
});
