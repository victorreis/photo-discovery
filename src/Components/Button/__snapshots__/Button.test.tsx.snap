// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component tests snapshot tests should render correctly 1`] = `
<button
  className="ButtonContainer-photo-discovery__sc-17i16aw-0 gVDkDM"
  data-testid="Button"
  onClick={[Function]}
  type="button"
>
  text
</button>
`;
